app/app.go

```diff
@@ -966,9 +966,15 @@ func (app *App) ProcessProposalHandler(ctx sdk.Context, req *abci.RequestProcess
 				optimisticProcessingInfo.Completion <- true
 			}()
 		}
-	} else if !bytes.Equal(app.optimisticProcessingInfo.Hash, req.Hash) {
-		app.optimisticProcessingInfo.Completion <- false
+	} else {
+		app.optimisticProcessingInfo.Completion = completionSignal
+		if !bytes.Equal(app.optimisticProcessingInfo.Hash, req.Hash) {
+			app.optimisticProcessingInfo.Completion <- false
+		} else {
+			app.optimisticProcessingInfo.Completion <- true
```

Comment:
we do not want to set true here since the OP goroutine could still be running at this time. It would cause ProcessBlock in FinalizeBlock to run in parallel with the OP goroutine which would cause non determinism